Makefile

```diff
@@ -1,21 +1,77 @@
+SHELL := bash
+.ONESHELL:
+.SHELLFLAGS := -eu -o pipefail -c
+.DELETE_ON_ERROR:
+MAKEFLAGS += --warn-undefined-variables
+MAKEFLAGS += --no-builtin-rules
+MAKEFLAGS += --no-print-directory
+MAKEFLAGS += -S
+
 .DEFAULT: all
 
+CARGO ?= cargo
+CARGO_WATCH ?= cargo-watch
+CARGO_CLIPPY ?= cargo-clippy
+DOCKER ?= docker
+
 all: build test
 
-build:
-	cargo build --features cli
+help:  ## Display this help
+	@awk 'BEGIN {FS = ":.*##"; printf "\nUsage:\n  make \033[36m<target>\033[0m\n"} /^[a-zA-Z_\-.*]+:.*?##/ { printf "  \033[36m%-15s\033[0m %s\n", $$1, $$2 } /^##@/ { printf "\n\033[1m%s\033[0m\n", substr($$0, 5) } ' $(MAKEFILE_LIST)
+
+###########
+# Tooling #
+###########
+
+# Ensure that cargo watch is installed
+check-cargo-watch:
+ifeq ("",$(shell command -v $(CARGO_WATCH)))
+	$(error "ERROR: cargo-watch is not installed (see: https://crates.io/crates/cargo-watch)")
+endif
+
+# Ensure that clippy is installed
+check-cargo-clippy:
+ifeq ("",$(shell command -v $(CARGO_CLIPPY)))
+	$(error "ERROR: clippy is not installed (see: https://doc.rust-lang.org/clippy/installation.html)")
+endif
+
+#########
+# Build #
+#########
 
+lint: check-cargo-clippy
+	$(CARGO) fmt --all --check
+	$(CARGO) clippy --all-features --all-targets --workspace
+
+build: ## Build wadm
+	$(CARGO) build
+
+build-watch: check-cargo-watch ## Build wadm continuously
+	$(CARGO) watch -- $(MAKE) build
+
+########
+# Test #
+########
+
+# An optional specific test for carog to target
+CARGO_TEST_TARGET ?=
+
+test:: ## Run tests
 ifeq ($(shell nc -czt -w1 127.0.0.1 4222 || echo fail),fail)
-test::
-	docker run --rm -d --name wadm-test -p 127.0.0.1:4222:4222 nats:2.9 -js
-	$(MAKE) test
-	docker stop wadm-test
+	$(DOCKER) run --rm -d --name wadm-test -p 127.0.0.1:4222:4222 nats:2.9 -js
+	$(CARGO) test $(CARGO_TEST_TARGET) -- --nocapture

```

Comment:
Why hard-code `--nocapture`? Why not stick with defaults and let the developer choose the arguments like one would do with a
```
cargo test ${@}
```
in a bash script

Reply:
While `$@` would work in a bash script, it doesn't work quite that way in `make` -- that's usually done with some sort of `ARGS` variable.

As for the `--nocapture` setting, I'm fine with removing it but this is basically what tests looked like across all of the wasmCloud codebase so far (probably to prevent surprise/skip a step when someone is `println!`-debugging a test). I'm fine with changing it, but was going for consistency/least surprise here